### PR TITLE
More explicit exception message when testing with fast_dev_run=True

### PR DESCRIPTION
**Pull request by hobogalaxy**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6615.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -948,6 +948,11 @@ def test(
         #  Attach dataloaders (if given)
         self.data_connector.attach_dataloaders(model, test_dataloaders=test_dataloaders)
 
+        if not model_provided and self.fast_dev_run:
+            raise MisconfigurationException(
+                f'You cannot execute testing when model is not provided and {self.fast_dev_run=}. Provide model with `trainer.test(model=...)`'
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:953]:
```suggestion
                f'You cannot execute testing when the model is not provided and {self.fast_dev_run}. Provide model with `trainer.test(model=...)`'
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -948,6 +948,12 @@ def test(
         #  Attach dataloaders (if given)
         self.data_connector.attach_dataloaders(model, test_dataloaders=test_dataloaders)
 
+        if not model_provided and ckpt_path == 'best' and self.fast_dev_run:
+            raise MisconfigurationException(
+                'You cannot execute testing when the model is not provided and `fast_dev_run=True`. '
+                'Provide model with `trainer.test(model=...)` or `trainer.test(ckpt_path=...)`'
+            )
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:956]: Please, move this into `__load_ckpt_weights`

That way, it is run also for `trainer.validate()`

**Review by carmocca — CHANGES_REQUESTED** [review]
See my comment.

Also, please update the CHANGELOG and add a test for the `MisconfigurationException`

Thanks!

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (hobogalaxy) [reply pytorch_lightning/trainer/trainer.py]: Done.

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -1777,3 +1777,19 @@ def on_fit_start(self, trainer, pl_module: LightningModule) -> None:
...
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, fast_dev_run=fast_dev_run)
+
+    trainer.fit(model)
+
+    if fast_dev_run:
+        with pytest.raises(MisconfigurationException, match=".*when `fast_dev_run=True`*"):
+            trainer.validate()
+
+        with pytest.raises(MisconfigurationException, match=".*when `fast_dev_run=True`*"):
+            trainer.test()
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1795]: No need to parametrize this test

```suggestion
def test_exception_when_testing_or_validating_with_fast_dev_run(tmpdir):
    model = BoringModel()
    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)

    trainer.fit(model)

    with pytest.raises(MisconfigurationException, match=".*when `fast_dev_run=True`*"):
        trainer.validate()

    with pytest.raises(MisconfigurationException, match=".*when `fast_dev_run=True`*"):
        trainer.test()
```

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -956,10 +956,17 @@ def __load_ckpt_weights(
...
         # if user requests the best checkpoint but we don't have it, error
-        if ckpt_path == 'best' and not self.checkpoint_callback.best_model_path:
-            raise MisconfigurationException(
-                'ckpt_path is "best", but `ModelCheckpoint` is not configured to save the best model.'
-            )
+        if ckpt_path == 'best':
+            if not self.checkpoint_callback.best_model_path and self.fast_dev_run:
+                raise MisconfigurationException(
+                    'You cannot execute `trainer.test()` or trainer.validate()`'
+                    ' when `fast_dev_run=True`.'
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:963]:
```suggestion
                    ' with `fast_dev_run=True`.'
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -956,10 +956,17 @@ def __load_ckpt_weights(
...
     ) -> Optional[str]:
         # if user requests the best checkpoint but we don't have it, error
-        if ckpt_path == 'best' and not self.checkpoint_callback.best_model_path:
-            raise MisconfigurationException(
-                'ckpt_path is "best", but `ModelCheckpoint` is not configured to save the best model.'
-            )
+        if ckpt_path == 'best':
+            if not self.checkpoint_callback.best_model_path and self.fast_dev_run:
+                raise MisconfigurationException(
+                    'You cannot execute `trainer.test()` or trainer.validate()`'
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:962]:
```suggestion
                    'You cannot execute `trainer.test()` or `trainer.validate()`'
```

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -956,10 +956,17 @@ def __load_ckpt_weights(
...
-            raise MisconfigurationException(
-                'ckpt_path is "best", but `ModelCheckpoint` is not configured to save the best model.'
-            )
+        if ckpt_path == 'best':
+            if not self.checkpoint_callback.best_model_path and self.fast_dev_run:
+                raise MisconfigurationException(
+                    'You cannot execute `trainer.test()` or `trainer.validate()`'
+                    ' with `fast_dev_run=True`.'
+                )
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:965]:
```suggestion
        if ckpt_path == 'best':
            if not self.checkpoint_callback.best_model_path and self.fast_dev_run:
                raise MisconfigurationException(
                    'You cannot execute `trainer.test()` or `trainer.validate()`'
                    ' with `fast_dev_run=True`.'
                )
                
                # if user requests the best checkpoint but we don't have it, error
```

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
